rust/operator-binary/src/controller.rs

```diff
@@ -0,0 +1,490 @@
+        // Serialize as a string to discourage users from trying to parse the value,
+        // and to keep things flexible if we end up changing the hasher at some point.
+        discovery_hash: Some(discovery_hash.finish().to_string()),
+    };
+    client
+        .apply_patch_status(FIELD_MANAGER_SCOPE, &hive, &status)
+        .await
+        .context(ApplyStatus)?;
+
+    Ok(ReconcilerAction {
+        requeue_after: None,
+    })
+}
+
+/// The server-role service is the primary endpoint that should be used by clients that do not perform internal load balancing,
+/// including targets outside of the cluster.
+///
+/// Note that you should generally *not* hard-code clients to use these services; instead, create a [`HiveZnode`](`stackable_zookeeper_crd::HiveZnode`)
+/// and use the connection string that it gives you.
+pub fn build_metastore_role_service(hive: &HiveCluster) -> Result<Service> {
+    let role_name = HiveRole::MetaStore.to_string();
+
+    let role_svc_name = hive
+        .metastore_role_service_name()
+        .context(GlobalServiceNameNotFound)?;
+    Ok(Service {
+        metadata: ObjectMetaBuilder::new()
+            .name_and_namespace(hive)
+            .name(&role_svc_name)
+            .ownerreference_from_resource(hive, None, Some(true))
+            .context(ObjectMissingMetadataForOwnerRef)?
+            .with_recommended_labels(hive, APP_NAME, hive_version(hive)?, &role_name, "global")
+            .build(),
+        spec: Some(ServiceSpec {
+            ports: Some(vec![ServicePort {
+                name: Some("hive".to_string()),
+                port: APP_PORT.into(),
+                protocol: Some("TCP".to_string()),
+                ..ServicePort::default()
+            }]),
+            selector: Some(role_selector_labels(hive, APP_NAME, &role_name)),
+            type_: Some("NodePort".to_string()),
+            ..ServiceSpec::default()
+        }),
+        status: None,
+    })
+}
+
+/// The rolegroup [`ConfigMap`] configures the rolegroup based on the configuration given by the administrator
+fn build_metastore_rolegroup_config_map(
+    rolegroup: &RoleGroupRef<HiveCluster>,
+    hive: &HiveCluster,
+    metastore_config: &HashMap<PropertyNameKind, BTreeMap<String, String>>,
+) -> Result<ConfigMap> {
+    let mut hive_site_data = String::new();
+    let log4j_data = include_str!("../../../deploy/external/log4j.properties");
+
+    for (property_name_kind, config) in metastore_config {
+        match property_name_kind {
+            PropertyNameKind::File(file_name) if file_name == HIVE_SITE_XML => {
+                let mut data = BTreeMap::new();
+
+                for (property_name, property_value) in config {
+                    data.insert(property_name.to_string(), Some(property_value.to_string()));
+                }
+
+                // TODO: make configurable
+                data.insert(
+                    "hive.metastore.warehouse.dir".to_string(),
```

Comment:
Use constant METASTORE_WAREHOUSE_DIR and I would remove the TODO.

Reply:
done